### PR TITLE
Added kernel goodness of fit

### LINKED ISSUE #103: Create a goodness of fit module in hyppo
http://papers.nips.cc/paper/6630-a-linear-time-kernel-goodness-of-fit-test
https://dl.acm.org/doi/10.5555/3305381.3305561

are two examples

### PR DESCRIPTION
**Pull request by GM-git-dotcom**
#### Reference issue
Close #103

#### Type of change
 Addition of functionality

#### What does this implement/fix?
 This will implement the kernel goodness of fit (Jitkrittum, Xu, Szabo, Fukumizu & Gretton 2017): https://proceedings.neurips.cc/paper/2017/file/979d472a84804b9f647bc185a877a8b5-Paper.pdf. A new module "kgof" is added in hyppo/tools, I am attaching a .txt file with the code
[kgof in hyppo demonstration.txt](https://github.com/neurodata/hyppo/files/7020056/kgof.in.hyppo.demonstration.txt)
 to demonstrate it. All credit goes to Jitkrittum et al. for the kernel goodness of fit.
Note: This comes with an added setup requirement for hyppo: autograd>=1.1.7 since it computes derivatives for optimization and replaces NumPy.
#### Additional information
 This is only the implementation of the first link given in #103 -- the second, (Jitkrittum, Szabo & Gretton 2017), is not implemented.

**Why I think CircleCI "run code linting" is failing for tools**: Since now there is a new requirement of autograd, I linted goftest.py (where the error originated) to find 2 errors:
E: 15, 0: Unable to import 'autograd' (import-error)
E: 16, 0: Unable to import 'autograd.numpy' (import-error)

But I've added autograd in the setup.py and requirements.txt. So I'm guessing CircleCI isn't still considering autograd a requirement.. or I'm missing another spot to specify autograd as the new requirement (apart from setup.py and requirements.txt). Runs fine locally.

### CODE REVIEW

**Comment by sampan501** [comment]
Hi thanks for the contribution. Unfortunately many of the changes revert new additions which I think may be because your forked branch has not been updated. Can you do this?

**Comment by sampan501** [comment]
> **Why I think CycleCI "run code linting" is failing for tools**: Since now there is a new requirement of autograd, I linted goftest.py (where the error originated) to find 2 errors:
> E: 15, 0: Unable to import 'autograd' (import-error)
> E: 16, 0: Unable to import 'autograd.numpy' (import-error)
> 
> But I've added autograd in the setup.py and requirements.txt. So I'm guessing CycleCI isn't still considering autograd a requirement.. or I'm missing another spot to specify autograd as the new requirement (apart from setup.py and requirements.txt). Runs fine locally.

CircleCI is likely using an old cache of dependencies. Since you added an additional one, you will have to increment the cache from `v2` to `v3` [here](https://github.com/neurodata/hyppo/blob/54c628f8141b9db24cd1edaf0847591f8fb2e4dd/.circleci/config.yml#L16) and everywhere else I use a `key` in that file.

Before doing that, I would see if you can modify your code so that `autograd` is not used, i.e. if the functionality is already in `numpy`. It seems the package hasn't been maintained for about 7 years and may cause a lot of dependency issues in the future.

**Comment by GM-git-dotcom** [comment]
I did pull the latest changes from master to my local repository before making changes using `git pull upstream master` after I forked it 2 days ago. The branch 'goodness_of_fit' now has a kgof folder in hyppo/tools. Can you elaborate on how the changes are reverting new additions?

Also, I agree, even I was reluctant on adding a new requirement for autograd initially. I will work on computing the derivatives using only numpy and let you know the updates.

**Comment by GM-git-dotcom** [comment]
I have removed autograd as dependency and wherever it was appearing in /tools/kgof, tested it locally and it checked out. However, after seeing the test fails in CircleCI I think there are now problems with /independence/dcorr.py and (still) with /tools/kgof/goftest.py. I linted them both and I'm not quite sure how to fix them -
```
pylint --errors-only dcorr.py
No config file found, using default configuration
************* Module hyppo.independence.dcorr
E:  4, 0: No name 'check_perm_blocks_dim' in module 'hyppo.tools' (no-name-in-module)
E:  4, 0: No name 'chi2_approx' in module 'hyppo.tools' (no-name-in-module)
E:  4, 0: No name 'compute_dist' in module 'hyppo.tools' (no-name-in-module)
E:314, 9: Instance of 'list' has no 'shape' member (no-member)
E:316,10: Instance of 'list' has no 'astype' member (no-member)
```

I think the errors on line 314 and 316 can easily be fixed by converting list type to an np array. However, I checked and the module names `check_perm_blocks_dim`, `chi2_approx`, and `compute_dist` exist as functions in /tools/common.py so I don't know why errors on line 4 are being thrown.

As for /tools/kgof/goftest.py -
```
pylint --errors-only goftest.py
No config file found, using default configuration
************* Module hyppo.tools.kgof.goftest
E:431,23: Module 'numpy.random' has no 'randn' member (no-member)
E:657,15: Module 'numpy.random' has no 'randint' member (no-member)
E:666, 8: Module 'numpy.random' has no 'multinomial' member (no-member)
```

This was rather strange as these are recognized members of numpy.random. I looked it up here as well (seems to be a minor issue?): https://stackoverflow.com/questions/20553551/how-do-i-get-pylint-to-recognize-numpy-members
Also, I'm quite new with CircleCI, please let me know if there is a way to see exactly what the errors are in 'run_code_linting' because as of now all I see is something like "x files need to be reformatted, exited with code exit status xxx" and so I lint these files locally.

**Comment by sampan501** [comment]
Ah I see. Looks like from the CircleCI error log that `black` formatting wasn't run on the code and so an error was thrown, so I would do that.

File: hyppo/discrim/discrim_one_samp.py [hunk 1]
```diff
@@ -1,21 +1,14 @@
-from typing import NamedTuple
-
 import numpy as np
 from scipy._lib._util import MapWrapper
 
 from ._utils import _CheckInputs
 from .base import DiscriminabilityTest
 
 
-class DiscrimOneSampleTestOutput(NamedTuple):
```
Comment (sampan501) [inline hyppo/discrim/discrim_one_samp.py:10]: > I did pull the latest changes from master to my local repository before making changes using git pull upstream master after I forked it 2 days ago. The branch 'goodness_of_fit' now has a kgof folder in hyppo/tools. Can you elaborate on how the changes are reverting new additions?

This is one example of what I was talking about. We added a `NamedTuple` for outputs recently, but those changes aren't included in the new file. Same with a few more changes in these files.

**Comment by sampan501** [comment]
One other thing I notice that is missing is testing your own unit tests. CircleCI should be able to do this by adding another module [here](https://github.com/neurodata/hyppo/blob/54c628f8141b9db24cd1edaf0847591f8fb2e4dd/.circleci/config.yml#L142) in a similar format to the other modules listed.

**Comment by sampan501** [comment]
Sorry for the barage of messages, but that last suggestion I had is minor. Generally, I like to reserve the `tools` module for common functions used by many of the modules, i.e. permutation test p-value computation, and local module specific functions within their modules. So, can you please move the `kgof` folder into a new module? Maybe called `gof`?

**Comment by GM-git-dotcom** [comment]
Thanks for the elaboration. I'll let you know when I'm done with these fixes.

**Pull request closed without merging**